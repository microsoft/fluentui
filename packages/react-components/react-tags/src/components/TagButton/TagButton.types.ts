import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type TagButtonSlots = {
  root: Slot<'div'>;
};

/**
 * TagButton Props
 */
export type TagButtonProps = ComponentProps<TagButtonSlots> & {};

/**
 * State used in rendering TagButton
 */
export type TagButtonState = ComponentState<TagButtonSlots>;
// TODO: Remove semicolon from previous line, uncomment next line, and provide union of props to pick from TagButtonProps.
// & Required<Pick<TagButtonProps, 'propName'>>
