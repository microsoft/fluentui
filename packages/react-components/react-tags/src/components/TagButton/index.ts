export * from './TagButton';
export * from './TagButton.types';
export * from './renderTagButton';
export * from './useTagButton';
export * from './useTagButtonStyles';
