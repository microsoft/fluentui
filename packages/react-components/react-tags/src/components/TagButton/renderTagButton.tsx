import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { TagButtonState, TagButtonSlots } from './TagButton.types';

/**
 * Render the final JSX of TagButton
 */
export const renderTagButton_unstable = (state: TagButtonState) => {
  const { slots, slotProps } = getSlots<TagButtonSlots>(state);

  // TODO Add additional slots in the appropriate place
  return <slots.root {...slotProps.root} />;
};
