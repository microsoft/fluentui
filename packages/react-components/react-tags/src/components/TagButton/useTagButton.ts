import * as React from 'react';
import { getNativeElementProps } from '@fluentui/react-utilities';
import type { TagButtonProps, TagButtonState } from './TagButton.types';

/**
 * Create the state required to render TagButton.
 *
 * The returned state can be modified with hooks such as useTagButtonStyles_unstable,
 * before being passed to renderTagButton_unstable.
 *
 * @param props - props from this instance of TagButton
 * @param ref - reference to root HTMLElement of TagButton
 */
export const useTagButton_unstable = (props: TagButtonProps, ref: React.Ref<HTMLElement>): TagButtonState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: getNativeElementProps('div', {
      ref,
      ...props,
    }),
  };
};
