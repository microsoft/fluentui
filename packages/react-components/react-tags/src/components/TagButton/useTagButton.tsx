import * as React from 'react';
import { getNativeElementProps, resolveShorthand, useEventCallback, useId } from '@fluentui/react-utilities';
import { DismissRegular, bundleIcon, DismissFilled } from '@fluentui/react-icons';
import type { TagButtonProps, TagButtonState } from './TagButton.types';
import { useARIAButtonShorthand } from '@fluentui/react-aria';
import { Delete, Backspace } from '@fluentui/keyboard-keys';
import { useTagGroupContext_unstable } from '../../contexts/TagGroupContext';

const tagButtonAvatarSizeMap = {
  medium: 28,
  small: 24,
  'extra-small': 20,
} as const;

const tagButtonAvatarShapeMap = {
  rounded: 'square',
  circular: 'circular',
} as const;

const DismissIcon = bundleIcon(DismissFilled, DismissRegular);

/**
 * Create the state required to render TagButton.
 *
 * The returned state can be modified with hooks such as useTagButtonStyles_unstable,
 * before being passed to renderTagButton_unstable.
 *
 * @param props - props from this instance of TagButton
 * @param ref - reference to root HTMLElement of TagButton
 */
export const useTagButton_unstable = (props: TagButtonProps, ref: React.Ref<HTMLElement>): TagButtonState => {
  const { dismissible: contextDismissible, handleTagDismiss, size: contextSize } = useTagGroupContext_unstable();

  const id = useId('fui-Tag', props.id);

  const {
    appearance = 'filled-lighter',
    disabled = false,
    dismissible = contextDismissible,
    shape = 'rounded',
    size = contextSize,
    value = id,
  } = props;

  const onDismissButtonClick = useEventCallback((ev: React.MouseEvent<HTMLButtonElement>) => {
    props.onClick?.(ev);
    if (!ev.defaultPrevented) {
      handleTagDismiss?.(ev, value);
    }
  });

  const onDismissButtonKeyDown = useEventCallback((ev: React.KeyboardEvent<HTMLButtonElement>) => {
    props?.onKeyDown?.(ev);
    if (!ev.defaultPrevented && (ev.key === Delete || ev.key === Backspace)) {
      handleTagDismiss?.(ev, value);
    }
  });

  const dismissButtonShorthand = resolveShorthand(props.dismissButton, {
    required: dismissible,
    defaultProps: {
      disabled,
      type: 'button',
      children: <DismissIcon />,
    },
  });

  return {
    appearance,
    avatarShape: tagButtonAvatarShapeMap[shape],
    avatarSize: tagButtonAvatarSizeMap[size],
    disabled,
    dismissible,
    shape,
    size,

    components: {
      root: 'div',
      content: 'div',
      media: 'span',
      icon: 'span',
      primaryText: 'span',
      secondaryText: 'span',
      dismissButton: 'button',
    },

    root: getNativeElementProps('div', {
      ref,
      ...props,
      id,
    }),

    content: useARIAButtonShorthand(props.content, {
      required: true,
      defaultProps: {
        disabled,
        tabIndex: 0,
        type: 'button',
      },
    }),
    media: resolveShorthand(props.media),
    icon: resolveShorthand(props.icon),
    primaryText: resolveShorthand(props.primaryText, { required: true }),
    secondaryText: resolveShorthand(props.secondaryText),

    dismissButton: {
      ...dismissButtonShorthand,
      onClick: onDismissButtonClick,
      onKeyDown: onDismissButtonKeyDown,
    },
  };
};
