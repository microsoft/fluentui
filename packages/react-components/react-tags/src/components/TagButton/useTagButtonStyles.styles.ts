import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { TagButtonSlots, TagButtonState } from './TagButton.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { createCustomFocusIndicatorStyle } from '@fluentui/react-tabster';
import { tokens } from '@fluentui/react-theme';
import {
  useIconStyles,
  useMediaStyles,
  usePrimaryTextStyles,
  useSecondaryTextStyles,
} from '../Tag/useTagStyles.styles';

export const tagButtonClassNames: SlotClassNames<TagButtonSlots> = {
  root: 'fui-TagButton',
  content: 'fui-TagButton__content',
  media: 'fui-TagButton__media',
  icon: 'fui-TagButton__icon',
  primaryText: 'fui-TagButton__primaryText',
  secondaryText: 'fui-TagButton__secondaryText',
  dismissButton: 'fui-TagButton__dismissButton',
};

const mediumIconSize = '20px';
const smallIconSize = '16px';
const extraSmallIconSize = '12px';

const useRootStyles = makeStyles({
  base: {
    display: 'inline-flex',
    alignItems: 'center',
    boxSizing: 'border-box',
    width: 'fit-content',

    backgroundColor: tokens.colorNeutralBackground3,
    color: tokens.colorNeutralForeground2,
    ...shorthands.border(tokens.strokeWidthThin, 'solid', tokens.colorTransparentStroke),
  },

  rounded: shorthands.borderRadius(tokens.borderRadiusMedium),
  circular: shorthands.borderRadius(tokens.borderRadiusCircular),

  medium: {
    height: '32px',
  },
  small: {
    height: '24px',
  },
  'extra-small': {
    height: '20px',
  },
});

const useContentButtonStyles = makeStyles({
  base: {
    // TODO use makeResetStyle when styles are settled

    // reset default button style:
    color: 'inherit',
    fontFamily: 'inherit',
    ...shorthands.padding(0),
    ...shorthands.borderStyle('none'),
    appearance: 'button',
    textAlign: 'unset',
    backgroundColor: 'transparent',

    display: 'inline-grid',
    height: '100%',
    alignItems: 'center',
    gridTemplateAreas: `
    "media primary  "
    "media secondary"
    `,

    ':hover': {
      cursor: 'pointer',
    },
  },

  rounded: createCustomFocusIndicatorStyle(
    {
      ...shorthands.borderRadius(tokens.borderRadiusMedium),
      ...shorthands.outline(tokens.strokeWidthThick, 'solid', tokens.colorStrokeFocus2),
      zIndex: 1,
    },
    { enableOutline: true },
  ),
  circular: createCustomFocusIndicatorStyle(
    {
      ...shorthands.borderRadius(tokens.borderRadiusCircular),
      ...shorthands.outline(tokens.strokeWidthThick, 'solid', tokens.colorStrokeFocus2),
      zIndex: 1,
    },
    { enableOutline: true },
  ),

  medium: {
    paddingRight: '7px',
  },
  small: {
    paddingRight: '5px',
  },
  'extra-small': {
    paddingRight: '5px',
  },
});
/**
 * Styles for content slot when TagButton is without leading media/icon
 */
const useContentButtonWithoutMediaStyles = makeStyles({
  medium: {
    paddingLeft: '7px',
  },
  small: {
    paddingLeft: '5px',
  },
  'extra-small': {
    paddingLeft: '5px',
  },
});
/**
 * Styles for content slot when TagButton has dismiss button
 */
const useDismissibleContentButtonStyles = makeStyles({
  base: createCustomFocusIndicatorStyle({
    borderTopRightRadius: tokens.borderRadiusNone,
    borderBottomRightRadius: tokens.borderRadiusNone,
  }),
  medium: {
    paddingRight: tokens.spacingHorizontalS,
  },
  small: {
    paddingRight: tokens.spacingHorizontalSNudge,
  },
  'extra-small': {
    paddingRight: tokens.spacingHorizontalSNudge,
  },
});

const useDismissButtonStyles = makeStyles({
  base: {
    // reset default button style:
    color: 'inherit',
    fontFamily: 'inherit',
    ...shorthands.padding(0),
    ...shorthands.borderStyle('none'),
    appearance: 'button',
    textAlign: 'unset',
    backgroundColor: 'transparent',

    display: 'flex',
    height: '100%',
    alignItems: 'center',

    // divider:
    ...shorthands.borderLeft(tokens.strokeWidthThin, 'solid', tokens.colorNeutralStroke1),
    borderTopLeftRadius: tokens.borderRadiusNone,
    borderBottomLeftRadius: tokens.borderRadiusNone,

    ':hover': {
      cursor: 'pointer',
    },
  },

  rounded: createCustomFocusIndicatorStyle({
    ...shorthands.outline(tokens.strokeWidthThick, 'solid', tokens.colorStrokeFocus2),
    borderTopLeftRadius: tokens.borderRadiusNone,
    borderBottomLeftRadius: tokens.borderRadiusNone,
    borderTopRightRadius: tokens.borderRadiusMedium,
    borderBottomRightRadius: tokens.borderRadiusMedium,
  }),
  circular: createCustomFocusIndicatorStyle({
    ...shorthands.outline(tokens.strokeWidthThick, 'solid', tokens.colorStrokeFocus2),
    borderTopLeftRadius: tokens.borderRadiusNone,
    borderBottomLeftRadius: tokens.borderRadiusNone,
    borderTopRightRadius: tokens.borderRadiusCircular,
    borderBottomRightRadius: tokens.borderRadiusCircular,
  }),

  medium: {
    fontSize: mediumIconSize,
    paddingLeft: '5px',
    paddingRight: '5px',
  },
  small: {
    fontSize: smallIconSize,
    paddingLeft: '3px',
    paddingRight: '3px',
  },
  'extra-small': {
    fontSize: extraSmallIconSize,
    paddingLeft: '5px',
    paddingRight: '5px',
  },
});

/**
 * Apply styling to the TagButton slots based on the state
 */
export const useTagButtonStyles_unstable = (state: TagButtonState): TagButtonState => {
  const rootStyles = useRootStyles();

  const contentButtonStyles = useContentButtonStyles();
  const contentButtonWithoutMediaStyles = useContentButtonWithoutMediaStyles();
  const dismissibleContentButtonStyles = useDismissibleContentButtonStyles();

  const iconStyles = useIconStyles();
  const mediaStyles = useMediaStyles();
  const primaryTextStyles = usePrimaryTextStyles();
  const secondaryTextStyles = useSecondaryTextStyles();
  const dismissButtonStyles = useDismissButtonStyles();

  const { shape, size } = state;

  state.root.className = mergeClasses(
    tagButtonClassNames.root,
    rootStyles.base,
    rootStyles[shape],
    rootStyles[size],
    state.root.className,
  );

  if (state.content) {
    state.content.className = mergeClasses(
      tagButtonClassNames.content,

      contentButtonStyles.base,
      contentButtonStyles[shape],
      contentButtonStyles[size],

      !state.media && !state.icon && contentButtonWithoutMediaStyles[size],
      state.dismissible && dismissibleContentButtonStyles.base,
      state.dismissible && dismissibleContentButtonStyles[size],

      state.content.className,
    );
  }

  if (state.media) {
    state.media.className = mergeClasses(
      tagButtonClassNames.media,
      mediaStyles.base,
      mediaStyles[size],
      state.media.className,
    );
  }
  if (state.icon) {
    state.icon.className = mergeClasses(
      tagButtonClassNames.icon,
      iconStyles.base,
      iconStyles[size],
      state.icon.className,
    );
  }
  if (state.primaryText) {
    state.primaryText.className = mergeClasses(
      tagButtonClassNames.primaryText,

      primaryTextStyles.base,
      primaryTextStyles[size],

      state.secondaryText ? primaryTextStyles.withSecondaryText : primaryTextStyles.withoutSecondaryText,

      state.primaryText.className,
    );
  }
  if (state.secondaryText) {
    state.secondaryText.className = mergeClasses(
      tagButtonClassNames.secondaryText,
      secondaryTextStyles.base,
      state.secondaryText.className,
    );
  }
  if (state.dismissButton) {
    state.dismissButton.className = mergeClasses(
      tagButtonClassNames.dismissButton,
      dismissButtonStyles.base,
      dismissButtonStyles[shape],
      dismissButtonStyles[size],
      state.dismissButton.className,
    );
  }

  return state;
};
