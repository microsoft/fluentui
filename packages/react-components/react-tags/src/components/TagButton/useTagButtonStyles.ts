import { makeStyles, mergeClasses } from '@griffel/react';
import type { TagButtonSlots, TagButtonState } from './TagButton.types';
import type { SlotClassNames } from '@fluentui/react-utilities';

export const tagButtonClassNames: SlotClassNames<TagButtonSlots> = {
  root: 'fui-TagButton',
  // TODO: add class names for all slots on TagButtonSlots.
  // Should be of the form `<slotName>: 'fui-TagButton__<slotName>`
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    // TODO Add default styles for the root element
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the TagButton slots based on the state
 */
export const useTagButtonStyles_unstable = (state: TagButtonState): TagButtonState => {
  const styles = useStyles();
  state.root.className = mergeClasses(tagButtonClassNames.root, styles.root, state.root.className);

  // TODO Add class names to slots, for example:
  // state.mySlot.className = mergeClasses(styles.mySlot, state.mySlot.className);

  return state;
};
