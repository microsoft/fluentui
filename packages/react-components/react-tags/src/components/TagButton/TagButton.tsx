import * as React from 'react';
import { useTagButton_unstable } from './useTagButton';
import { renderTagButton_unstable } from './renderTagButton';
import { useTagButtonStyles_unstable } from './useTagButtonStyles';
import type { TagButtonProps } from './TagButton.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * TagButton component - TODO: add more docs
 */
export const TagButton: ForwardRefComponent<TagButtonProps> = React.forwardRef((props, ref) => {
  const state = useTagButton_unstable(props, ref);

  useTagButtonStyles_unstable(state);
  return renderTagButton_unstable(state);
});

TagButton.displayName = 'TagButton';
