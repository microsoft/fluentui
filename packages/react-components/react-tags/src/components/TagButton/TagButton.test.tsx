import * as React from 'react';
import { render } from '@testing-library/react';
import { TagButton } from './TagButton';
import { isConformant } from '../../testing/isConformant';

describe('TagButton', () => {
  isConformant({
    Component: TagButton,
    displayName: 'TagButton',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<TagButton>Default TagButton</TagButton>);
    expect(result.container).toMatchSnapshot();
  });
});
