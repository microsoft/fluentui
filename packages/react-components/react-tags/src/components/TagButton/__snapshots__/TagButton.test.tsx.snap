// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagButton renders a default state 1`] = `
<div>
  <div
    class="fui-TagButton"
  >
    Default TagButton
  </div>
</div>
`;
